### PR TITLE
Fix coinbase height deserialization

### LINKED ISSUE #2226: Require minimal height encodings during deserialization
## Motivation

Zebra accepts non-minimal height encodings when deserializing, but the spec requires minimal encodings.

The serialization should be correct. But the deserialization accepts some encodings that are longer than they need to be. So they are actually invalid.

### Consensus Rules

Here's the relevant consensus rule:

> A coinbase transaction for a block at block height greater than 0 MUST have a script that, as its first item, encodes the block height height as follows.
> 
> For height in the range {1 .. 16}, the encoding is a single byte of value 0x50 + height.
> 
> Otherwise, let heightBytes be the signed little-endian representation of height, using the minimum nonzero number of bytes such that the most significant byte is < 0x80. The length of heightBytes MUST be in the range {1 .. 5}. Then the encoding is the length of heightBytes encoded as one byte, followed by heightBytes itself.
> 
> This matches the encoding used by Bitcoin in the implementation of [BIP-34] (but the description here is to be considered normative).

https://zips.z.cash/protocol/protocol.pdf#txnconsensus

### Related Consensus Rules

Here is a related consensus rule that limits the number of bytes used by a height:

> Each block in a block chain has a block height . The block height of the genesis block is 0, and the block height of each subsequent block in the block chain increments by 1. Implementations MUST support block heights up to and including 2^31 − 1.
> 
> As of NU5, there is a consensus rule that all coinbase transactions MUST have the nExpiryHeight field set to the block height , and this limits the maximum block height to 2^32 − 1, absent future consensus changes.

https://zips.z.cash/protocol/protocol.pdf#blockchain

## Existing Code

https://github.com/ZcashFoundation/zebra/blob/24bf952e982bde28eb384b211659159d46150f63/zebra-chain/src/transparent/serialize.rs#L44-L53

## Tests

We don't have any proptests for height bytes round-trips (bytes to `Height` to bytes).

There are two different encodings we need to test:
- [ ] single byte opcodes
- [ ] 1 length byte followed by 1-5 bytes of height data

We already have height deserialization tests for every mainnet and testnet block test vector.

## Context

Comment by @daira in https://github.com/ZcashFoundation/zebra/issues/2225#issuecomment-851049413

Support the large block heights required by the spec #1113 - out of scope for NU5

### PR DESCRIPTION
**Pull request by oxarbitrage**
## Motivation

In height deserialization we are allowing more than the spec. This will close https://github.com/ZcashFoundation/zebra/issues/2226 if merged.

### Specifications

> A coinbase transaction for a block at block height greater than 0 MUST have a script that, as its first item, encodes the block height height as follows.
>
> For height in the range {1 .. 16}, the encoding is a single byte of value 0x50 + height.
>
> Otherwise, let heightBytes be the signed little-endian representation of height, using the minimum nonzero number of bytes such that the most significant byte is < 0x80. The length of heightBytes MUST be in the range {1 .. 5}. Then the encoding is the length of heightBytes encoded as one byte, followed by heightBytes itself.
>
>This matches the encoding used by Bitcoin in the implementation of [BIP-34] (but the description here is to be considered normative).

## Solution

Constrain the decoding by block height ranges. Added some tests. 

## Review

@teor2345 is probably the most familiar with this but anyone can take a look.

### Reviewer Checklist

  - [ ] Code implements Specs and Designs
  - [ ] Tests for Expected Behaviour
  - [ ] Tests for Errors

### CODE REVIEW

File: zebra-chain/src/transparent/serialize.rs [hunk 1]
```diff
@@ -78,20 +78,33 @@ fn parse_coinbase_height(
         // Blocks 17 through 128 exclusive encode block height with the `0x01` opcode.
         // The Bitcoin encoding requires that the most significant byte is below 0x80.
         (Some(0x01), len) if len >= 2 && data[1] < 0x80 => {
-            Ok((Height(data[1] as u32), CoinbaseData(data.split_off(2))))
+            let h = data[1] as u32;
+            if h < 128 {
```
Comment (teor2345) [inline zebra-chain/src/transparent/serialize.rs]: Likely Bug: 1..16 are covered by the shorter encoding above, so they are not allowed in this longer encoding.

```suggestion
            if (17..128).contains(&h) {
```

File: zebra-chain/src/transparent/tests/vectors.rs [hunk 2]
```diff
@@ -0,0 +1,19 @@
+use super::super::serialize::parse_coinbase_height;
+
+#[test]
+fn parse_coinbase_height_mins() {
```
Comment (teor2345) [inline zebra-chain/src/transparent/tests/vectors.rs:4]: Test Coverage: Can you please the height `1` examples from the ticket as well?

We need to cover heights `1..=16`, which have a one-byte encoding.

File: zebra-chain/src/transparent/tests/prop.rs [hunk 3]
```diff
@@ -45,3 +46,25 @@ fn input_coinbase_vecs_only_have_coinbase_input() -> Result<()> {
 
     Ok(())
 }
+
+#[test]
+fn coinbase_height_round_trip() -> Result<()> {
```
Comment (teor2345) [inline zebra-chain/src/transparent/tests/prop.rs]: Test Coverage: Can you please add a byte->height->byte round-trip test?

We need to start with bytes, so we can find decoding bugs that allow multiple encodings for a single height.

Here's a similar test:
https://github.com/ZcashFoundation/zebra/blob/b1acdc61b4791baed5909809e860099ef6782597/zebra-network/src/protocol/external/tests/prop.rs#L46

**Review by teor2345 — COMMENTED** [review]
Looks good, but I think a few more tests would help.

File: zebra-chain/src/transparent/serialize.rs (hunk 1 above)
Comment (oxarbitrage) [reply zebra-chain/src/transparent/serialize.rs]: Thank you for this one it seems i missed it. Fixed it in https://github.com/ZcashFoundation/zebra/pull/3129/commits/8c2064526a8858554ec3f91ae8297012f76fa78b

File: zebra-chain/src/transparent/tests/vectors.rs (hunk 2 above)
Comment (oxarbitrage) [reply zebra-chain/src/transparent/tests/vectors.rs:4]: Sure, added them at https://github.com/ZcashFoundation/zebra/pull/3129/commits/73907af994d587b420c653b659fa58d4113cb9d0

File: zebra-chain/src/transparent/tests/prop.rs (hunk 3 above)
Comment (oxarbitrage) [reply zebra-chain/src/transparent/tests/prop.rs]: https://github.com/ZcashFoundation/zebra/pull/3129/commits/a40815d236d6bab5053737b0c73f3c55b44a40e1

File: zebra-chain/src/transparent/tests/vectors.rs [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+    zebra_test::init();
+
+    // examples with height 1:
+
+    let case1 = vec![0x51];
+    assert!(!parse_coinbase_height(case1.clone()).is_err());
+    assert_eq!(parse_coinbase_height(case1).unwrap().0 .0, 1);
```
Comment (teor2345) [inline zebra-chain/src/transparent/tests/vectors.rs:11]: Nitpick: Extra space, possibly confusing

```suggestion
    assert_eq!(parse_coinbase_height(case1).unwrap().0.0, 1);
```

File: zebra-chain/src/transparent/tests/vectors.rs [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+    assert!(parse_coinbase_height(case4).is_err());
+
+    let case5 = vec![0x04, 0x01, 0x00, 0x00, 0x00];
+    assert!(parse_coinbase_height(case5).is_err());
+
+    // examples with height 17:
+
+    let case1 = vec![0x01, 0x11];
+    assert!(!parse_coinbase_height(case1.clone()).is_err());
+    assert_eq!(parse_coinbase_height(case1).unwrap().0 .0, 17);
```
Comment (teor2345) [inline zebra-chain/src/transparent/tests/vectors.rs:29]: Nitpick: Extra space, possibly confusing

```suggestion
    assert_eq!(parse_coinbase_height(case1).unwrap().0.0, 17);
```

File: zebra-chain/src/transparent/tests/prop.rs [hunk 6]
```diff
@@ -0,0 +1,88 @@
...
+
+        prop_assert_eq!(height, decoded.0);
+    });
+
+    Ok(())
+}
+
+proptest! {
+    #[test]
+    fn coinbase_height_round_trip_from_random_bytes(mut height_bytes in vec(any::<u8>(), 1..4)) {
```
Comment (teor2345) [inline zebra-chain/src/transparent/tests/prop.rs]: Test Coverage: Length-prefixed heights can be up to 5 bytes long:

> ... The length of heightBytes MUST be in the range {1 .. 5}. Then the encoding is the length of heightBytes encoded as one byte, followed by heightBytes itself.

```suggestion
    fn coinbase_height_round_trip_from_random_bytes(mut height_bytes in vec(any::<u8>(), 1..5)) {
```

File: zebra-chain/src/transparent/tests/prop.rs [hunk 7]
```diff
@@ -0,0 +1,88 @@
...
+        let mut encoded1 = vec![height_bytes.len() as u8];
```
Comment (teor2345) [inline zebra-chain/src/transparent/tests/prop.rs:77]: Test Coverage: Can you also add a test for the one-byte encoding?

> For height in the range {1 .. 16}, the encoding is a single byte ...

**Review by teor2345 — COMMENTED** [review]
Almost there, we just need to cover all the possible encoded byte formats in the tests.

File: zebra-chain/src/transparent/tests/vectors.rs (hunk 4 above)
Comment (oxarbitrage) [reply zebra-chain/src/transparent/tests/vectors.rs:11]: Actually for some reason rustfmt is doing this :(

Comment (teor2345) [reply zebra-chain/src/transparent/tests/vectors.rs:11]: Oh wow, yeah, then let's just leave it.

Maybe it's worth reporting a rustfmt bug, but that's totally optional:
https://github.com/rust-lang/rustfmt/issues/new

Comment (oxarbitrage) [reply zebra-chain/src/transparent/tests/vectors.rs:11]: I opened an issue https://github.com/rust-lang/rustfmt/issues/5123

Not sure if it is a bug, i think there should be a reason. I think that if you have to do something as `var.0.0` it should mean that the involved function and types are not the best way designed, however i do not agree with rustfmt on adding an extra space to gain some clarity initially.

**Review by teor2345 — APPROVED** [review]
Looks great, thanks for writing all those tests!

**Pull request merged**